kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/EndOffsetsPoller.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2023 Responsive Computing, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.clients;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.serialization.BytesDeserializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class polls the current end offset of the topic partitions assigned locally and
+ * publishes the values as a metric. It gets the set of assigned partitions by listening
+ * on ResponsiveConsumer.
+ *
+ */
+public class EndOffsetsPoller {
+  private static final Logger LOGGER = LoggerFactory.getLogger(EndOffsetsPoller.class);
+
+  private final Map<String, ThreadMetrics> threadIdToMetrics = new HashMap<>();
+  private final Metrics metrics;
+  private final Map<String, Object> configs;
+  private final Factories factories;
+  private ScheduledExecutorService executor;
+  private ScheduledFuture<?> pollFuture;
+  private Consumer<?, ?> consumer;
```

Comment:
If this is only ever used to poll the end offsets, why not just use an admin client?

Reply:
I swear that was my original plan, but I didn't see the `listOffsets` method for some reason. I'll change it.